pkg/kubevirt/core/core.go

```diff
@@ -0,0 +1,300 @@
+						},
+						Resources: kubevirtv1.ResourceRequirements{
+							Requests: *requestsAndLimits,
+							Limits:   *requestsAndLimits,
+						},
+					},
+					TerminationGracePeriodSeconds: &terminationGracePeriodSeconds,
+					Volumes: []kubevirtv1.Volume{
+						{
+							Name: "datavolumedisk",
+							VolumeSource: kubevirtv1.VolumeSource{
+								DataVolume: &kubevirtv1.DataVolumeSource{
+									Name: machineName,
+								},
+							},
+						},
+						{
+							Name: "cloudinitdisk",
+							VolumeSource: kubevirtv1.VolumeSource{
+								CloudInitNoCloud: &kubevirtv1.CloudInitNoCloudSource{
+									UserDataSecretRef: &corev1.LocalObjectReference{
+										Name: userdataSecretName,
+									},
+								},
+							},
+						},
+					},
+					DNSPolicy: dnsPolicy,
+					DNSConfig: dnsConfig,
+				},
+			},
+			DataVolumeTemplates: []cdi.DataVolume{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: machineName,
+					},
+					Spec: cdi.DataVolumeSpec{
+						PVC: &corev1.PersistentVolumeClaimSpec{
+							StorageClassName: utilpointer.StringPtr(providerSpec.StorageClassName),
+							AccessModes: []corev1.PersistentVolumeAccessMode{
+								"ReadWriteOnce",
+							},
+							Resources: corev1.ResourceRequirements{
+								Requests: pvcRequest,
+							},
+						},
+						Source: cdi.DataVolumeSource{
+							HTTP: &cdi.DataVolumeSourceHTTP{
+								URL: providerSpec.SourceURL,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	kubevirtClient, err := kubevirtClient(secrets)
+	if err != nil {
+		return "", fmt.Errorf("failed to get kubevirt client: %v", err)
+	}
+
+	if err := kubevirtClient.Create(ctx, virtualMachine); err != nil {
+		return "", fmt.Errorf("failed to create vmi: %v", err)
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            userdataSecretName,
+			Namespace:       virtualMachine.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(virtualMachine, kubevirtv1.VirtualMachineGroupVersionKind)},
+		},
+		Data: map[string][]byte{"userdata": []byte(secrets.Data["userData"])},
+	}
+
+	if err := kubevirtClient.Create(ctx, secret); err != nil {
+		return "", fmt.Errorf("failed to create secret for userdata: %v", err)
+	}
+
+	// TODO(MQ): do we really need this? another approach is to return an empty provider id and leave it to the
+	// Get
+	return machineProviderID(secrets, machineName, providerSpec.Namespace)
+}
+
+func (p PluginSPIImpl) DeleteMachine(ctx context.Context, machineName, providerID string, providerSpec *api.KubeVirtProviderSpec, secrets *corev1.Secret) (foundProviderID string, err error) {
+	vm, err := getVM(secrets, machineName, providerSpec.Namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if vm != nil {
+		kubevirtClient, err := kubevirtClient(secrets)
+		if err != nil {
+			return "", fmt.Errorf("failed to create kubevirt client: %v", err)
+		}
+
+		if err := kubevirtClient.Delete(context.Background(), vm); err != nil {
```

Comment:
nonfound error should be ignored in this case.

Reply:
Can you elaborate more about the meaning of "nonfound" error?

Reply:
https://github.com/kubernetes/apimachinery/blob/7e441e0f246a2db6cf1855e4110892d1623a80cf/pkg/api/errors/errors.go#L478-L482